include/dlaf/factorization/qr/t_factor_impl.h

```diff
@@ -147,53 +211,117 @@ struct Helpers<Backend::MC, Device::CPU, T> {
                          });
   }
-
-  static auto stepGEMV(matrix::ReadOnlyTileSender<T, Device::GPU> tile_vi,
-                       matrix::ReadOnlyTileSender<T, Device::CPU> taus,
-                       matrix::ReadWriteTileSender<T, Device::GPU> tile_t) noexcept {
-    auto gemv_func = [](cublasHandle_t handle, const matrix::Tile<const T, Device::GPU>& tile_v,
+  static void loop_gemv(cublasHandle_t handle, const matrix::Tile<const T, Device::GPU>& tile_v,
                         const matrix::Tile<const T, Device::CPU>& taus,
-                        matrix::Tile<T, Device::GPU>& tile_t) noexcept {
-      const SizeType m = tile_v.size().rows();
-      const SizeType k = tile_t.size().cols();
-      DLAF_ASSERT(tile_v.size().cols() == k, tile_v.size().cols(), k);
-      DLAF_ASSERT(taus.size().rows() == k, taus.size().rows(), k);
-      DLAF_ASSERT(taus.size().cols() == 1, taus.size().cols());
-
-      gpulapack::larft_gemv0(handle, m, k, tile_v.ptr(), tile_v.ld(), taus.ptr(), tile_t.ptr(),
-                             tile_t.ld());
+                        const matrix::Tile<T, Device::GPU>& tile_t) noexcept {
+    const SizeType m = tile_v.size().rows();
+    const SizeType k = tile_t.size().cols();
+    DLAF_ASSERT(tile_v.size().cols() == k, tile_v.size().cols(), k);
+    DLAF_ASSERT(taus.size().rows() == k, taus.size().rows(), k);
+    DLAF_ASSERT(taus.size().cols() == 1, taus.size().cols());
 
-      return std::move(tile_t);
-    };
+    gpulapack::larft_gemv0(handle, m, k, tile_v.ptr(), tile_v.ld(), taus.ptr(), tile_t.ptr(),
+                           tile_t.ld());
+  }
 
+  static matrix::ReadWriteTileSender<T, Device::GPU> step_gemv(
+      matrix::Panel<Coord::Col, T, Device::GPU>& hh_panel,
+      matrix::ReadOnlyTileSender<T, Device::CPU> taus,
+      matrix::ReadWriteTileSender<T, Device::GPU> tile_t,
+      std::vector<matrix::ReadWriteTileSender<T, Device::GPU>> workspaces) {
     namespace ex = pika::execution::experimental;
     namespace di = dlaf::internal;
 
-    return ex::when_all(std::move(tile_vi), std::move(taus), std::move(tile_t)) |
-           di::transform<di::TransformDispatchType::Blas>(
-               di::Policy<Backend::GPU>(pika::execution::thread_priority::high), std::move(gemv_func));
+    using pika::execution::thread_priority;
+
+    std::vector<matrix::ReadOnlyTileSender<T, Device::GPU>> hh_tiles =
+        selectRead(hh_panel, hh_panel.iteratorLocal());
+
+    const auto workers_params = split_tfactor_work(hh_tiles.size());
+    const std::size_t nworkers = workers_params.nworkers;
+    const std::size_t batch_size = workers_params.batch_size;
+    DLAF_ASSERT(workspaces.size() >= nworkers - 1, workspaces.size(), nworkers - 1);
+
+    for (std::size_t worker_id = 0; worker_id < nworkers; ++worker_id) {
+      const std::size_t begin = worker_id * batch_size;
+      const std::size_t end = std::min(hh_tiles.size(), (worker_id + 1) * batch_size);
+
+      if (end - begin <= 0)
```

Comment:
Are you expecting to cover cases where `end < begin` as well? If yes, perhaps do
```suggestion
      if (end <= begin)
```
since they're unsigned integers. If not, perhaps explicitly do
```suggestion
      if (end - begin == 0)
```
instead to make it clear that the value can't be negative? Though I think I prefer `end <= begin` in any case since it'll keep working with signed and unsigned integers.

Reply:
AFAICT it should never happen `end < begin`, but the logic is not straightforward so it might be worth adding some check.

If you look just below, you can see that there is an implicit assumption about this

https://github.com/eth-cscs/DLA-Future/blob/0d2bd4f41f485f249f1c139898914823cedc2ed8/include/dlaf/factorization/qr/t_factor_impl.h#L262

Reply:
Sorry, I missed my conclusion:

- I like `end == begin`
- I would probably add a moderate/heavy assert with `end >= begin`

what would you go for?

Reply:
Thanks for the update (https://github.com/eth-cscs/DLA-Future/pull/1214/commits/37bedc3ff3820ecce1ec88de3b16e776b7c82ce0)! If you now go for the `end == begin` version, I would still add an assertion that `end >= begin` as well. If it's not true, things would blow up spectacularly later.

Reply:
Sorry, I posted my above comment without seeing your new comment! I think I already answered your question, but to be explicit: I would add the assert.

Making the condition for the if-clause `end <= begin` only makes sense if you actually expect `end` to sometimes be strictly smaller than `begin`, but if I understand correctly, you don't expect that to ever happen.

Reply:
see https://github.com/eth-cscs/DLA-Future/pull/1214/commits/021c642f7c6260e2a5e58476f6faa2690613900e and https://github.com/eth-cscs/DLA-Future/pull/1214/commits/2eae9d5cf37244cf8604d6fe76c592b4e0c1c10b